server/src/main/java/org/opensearch/search/wlm/tracker/QueryGroupResourceUsageTrackerService.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.wlm.tracker;
+
+import org.opensearch.search.ResourceType;
+import org.opensearch.search.wlm.QueryGroupHelper;
+import org.opensearch.search.wlm.QueryGroupLevelResourceUsageView;
+import org.opensearch.search.wlm.QueryGroupTask;
+import org.opensearch.tasks.Task;
+import org.opensearch.tasks.TaskManager;
+import org.opensearch.tasks.TaskResourceTrackingService;
+
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * This class tracks resource usage per QueryGroup
+ */
+// @ExperimentalApi
+public class QueryGroupResourceUsageTrackerService implements QueryGroupUsageTracker, TaskManager.TaskEventListeners {
+
+    public static final List<ResourceType> TRACKED_RESOURCES = List.of(ResourceType.fromName("memory"), ResourceType.fromName("cpu"));
+    private final TaskManager taskManager;
+    private final TaskResourceTrackingService taskResourceTrackingService;
+
+    /**
+     * QueryGroupResourceTrackerService constructor
+     *
+     * @param taskManager Task Manager service for keeping track of currently running tasks on the nodes
+     * @param taskResourceTrackingService Service that helps track resource usage of tasks running on a node.
+     */
+    public QueryGroupResourceUsageTrackerService(
+        final TaskManager taskManager,
+        final TaskResourceTrackingService taskResourceTrackingService
+    ) {
+        this.taskManager = taskManager;
+        this.taskResourceTrackingService = taskResourceTrackingService;
+    }
+
+    /**
+     * Constructs a map of QueryGroupLevelResourceUsageView instances for each QueryGroup.
+     *
+     * @return Map of QueryGroup views
+     */
+    @Override
+    public Map<String, QueryGroupLevelResourceUsageView> constructQueryGroupLevelUsageViews() {
+        Map<String, QueryGroupLevelResourceUsageView> queryGroupViews = new HashMap<>();
+
+        Map<String, List<Task>> tasksByQueryGroup = getTasksGroupedByQueryGroup();
+        Map<String, EnumMap<ResourceType, Long>> queryGroupResourceUsage = getResourceUsageOfQueryGroups(tasksByQueryGroup);
+
+        for (String queryGroupId : tasksByQueryGroup.keySet()) {
+            QueryGroupLevelResourceUsageView queryGroupLevelResourceUsageView = new QueryGroupLevelResourceUsageView(
+                queryGroupId,
+                queryGroupResourceUsage.get(queryGroupId),
+                tasksByQueryGroup.get(queryGroupId)
+            );
+            queryGroupViews.put(queryGroupId, queryGroupLevelResourceUsageView);
+        }
```

Comment:
Why not create the queryGroupView within the resourceUsage method? Or inline that method here?

Reply:
having `getResourceUsageOfQueryGroups` makes `constructQueryGroupLevelUsageViews` more readable with lesser lines and an opportunity to `getResourceUsageOfQueryGroups` from other methods in the future. 
If we agree on that, `constructQueryGroupLevelUsageViews` main job is to create these queryGroupLevelUsageViews and hence I'm keeping that in the same method and not in the `constructQueryGroupLevelUsageViews`

Reply:
How about below:

```
final Map<String, List<Task>> tasksByQueryGroup = getTasksGroupedByQueryGroup();
final Map<String, QueryGroupLevelResourceUsageView> queryGroupViews = new HashMap<>();

// Iterate over each QueryGroup entry
for (Map.Entry<String, List<Task>> queryGroupEntry : tasksByQueryGroup.entrySet()) {
    // Compute the QueryGroup usage
    final EnumMap<ResourceType, Long> queryGroupUsage = resourceUsageOfQueryGroups.get(queryGroupEntry.getKey());
    for (ResourceType resourceType : TRACKED_RESOURCES) {
        long queryGroupResourceUsage = 0;
        for (Task task : queryGroupEntry.getValue()) {
            queryGroupResourceUsage += QueryGroupHelper.getResourceUsage(resourceType, task);
        }
        queryGroupUsage.put(resourceType, queryGroupResourceUsage);
    }

    // Add to the QueryGroup View
    queryGroupViews.put(queryGroupEntry.getKey(), new QueryGroupLevelResourceUsageView(
                queryGroupEntry.getKey(),
                queryGroupUsage,
                queryGroupEntry.getValue());
}
return queryGroupViews;
```